code/go/0chain.net/core/util/mpt_pnodedb_test.go

```diff
@@ -0,0 +1,339 @@
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name:    "Test_NewPNodeDB_OK",
+			args:    args{dataDir: dataDir},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewPNodeDB(tt.args.dataDir, tt.args.logDir)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewPNodeDB() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+
+			if err := cleanUp(); err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestPNodeDB_Iterate(t *testing.T) {
+	db, cleanUp := newPNodeDB(t)
+	defer cleanUp()
+
+	wo := gorocksdb.NewDefaultWriteOptions()
+	err := db.db.Put(wo, []byte("key"), make([]byte, 0))
+	require.NoError(t, err)
+
+	type fields struct {
+		dataDir  string
+		db       *gorocksdb.DB
+		ro       *gorocksdb.ReadOptions
+		wo       *gorocksdb.WriteOptions
+		to       *gorocksdb.TransactionOptions
+		fo       *gorocksdb.FlushOptions
+		version  int64
+		versions []int64
+	}
+	type args struct {
+		ctx     context.Context
+		handler NodeDBIteratorHandler
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		wantErr bool
+	}{
+		{
+			name:    "Test_PNodeDB_Iterate_Err_Reading_OK",
+			fields:  fields{db: db.db},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			pndb := &PNodeDB{
+				dataDir:  tt.fields.dataDir,
+				db:       tt.fields.db,
+				ro:       tt.fields.ro,
+				wo:       tt.fields.wo,
+				to:       tt.fields.to,
+				fo:       tt.fields.fo,
+				version:  tt.fields.version,
+				versions: tt.fields.versions,
+			}
+			if err := pndb.Iterate(tt.args.ctx, tt.args.handler); (err != nil) != tt.wantErr {
+				t.Errorf("Iterate() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestPNodeDB_PruneBelowVersion_Iterator_Err(t *testing.T) {
+	db, cleanUp := newPNodeDB(t)
+	defer cleanUp()
+
+	wo := gorocksdb.NewDefaultWriteOptions()
+
+	for i := 0; i < 257; i++ {
+		r := rand.Int()
+		err := db.db.Put(wo, []byte(strconv.Itoa(r)), []byte{NodeTypeValueNode})
+		require.NoError(t, err)
+	}
+
+	wo.DisableWAL(true)
+	wo.SetSync(true)
+
+	fo := gorocksdb.NewDefaultFlushOptions()
+	ctx := context.WithValue(context.TODO(), PruneStatsKey, &PruneStats{})
+
+	type fields struct {
+		dataDir  string
+		db       *gorocksdb.DB
+		ro       *gorocksdb.ReadOptions
+		wo       *gorocksdb.WriteOptions
+		to       *gorocksdb.TransactionOptions
+		fo       *gorocksdb.FlushOptions
+		version  int64
+		versions []int64
+	}
+	type args struct {
+		ctx     context.Context
+		version Sequence
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		wantErr bool
+	}{
+		{
+			name:    "Test_PNodeDB_PruneBelowVersion_OK",
+			fields:  fields{db: db.db, wo: wo, fo: fo},
+			args:    args{ctx: ctx, version: 1},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			pndb := &PNodeDB{
+				dataDir:  tt.fields.dataDir,
```

Comment:
Move the fields defention inside the t.Run goroutine. The database needs to be setup inside the `t.Run` goroutine to avoid errors when used with `t.Parallel`.  For example
```go
for _, tt := range tests {
    t.Run(tt.name, func(t *testing.T) {
        t.Parallel()
        db, cleanUp := newPNodeDB(t)
        defer cleanUp()
        wo := gorocksdb.NewDefaultWriteOptions()
        for i := 0; i < 257; i++ {
            r := rand.Int()
            err := db.db.Put(wo, []byte(strconv.Itoa(r)), []byte{NodeTypeValueNode})
            require.NoError(t, err)
        }
        wo.DisableWAL(true)
        wo.SetSync(true)
        fo := gorocksdb.NewDefaultFlushOptions()
        ctx := context.WithValue(context.TODO(), PruneStatsKey, &PruneStats{})
        pndb := &PNodeDB{
            dataDir:  tt.fields.dataDir,
            db:       db.db,
            ro:       tt.fields.ro,
            wo:       wo,
            to:       tt.fields.to,
            fo:       fo,
            version:  tt.fields.version,
            versions: tt.fields.versions,
        }
        if err := pndb.PruneBelowVersion(ctx, tt.args.version); (err != nil) != tt.wantErr {
            t.Errorf("PruneBelowVersion() error = %v, wantErr %v", err, tt.wantErr)
        }
    })
}
```

Reply:
Done.